flux-refineck/src/constraint_gen.rs

```diff
@@ -539,8 +570,70 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     self.subtyping(rcx, ty1, ty2);
                 }
             }
+            (BaseTy::Generator(def_id, args), BaseTy::Alias(rty::AliasKind::Opaque, alias_ty)) => {
+                // TODO: use alias_ty.args
+                let obligs = mk_generator_obligations(
+                    self.genv,
+                    def_id,
+                    args,
+                    &alias_ty.def_id,
+                    self.span(),
+                );
+                match obligs {
+                    Ok(obligs) => {
+                        self.insert_obligations(obligs);
+                    }
+                    Err(e) => {
+                        panic!("unexpected error: {:#?}", e);
+                    }
+                }
+            }
+            (_, BaseTy::Alias(rty::AliasKind::Opaque, alias_ty)) => {
+                // TODO: use alias_ty.args
+                self.opaque_subtyping(rcx, bty1, alias_ty.def_id);
+            }
             _ => {
-                tracked_span_bug!("unexpected base types: `{:?}` and `{:?}`", bty1, bty2,);
+                panic!("unexpected base types: `{:?}` and `{:?}`", bty1, bty2,);
+            }
+        }
+    }
+
+    fn opaque_predicates(
+        &mut self,
+        opaque_def_id: DefId,
+        span: Span,
+    ) -> Option<Vec<rty::ProjectionPredicate>> {
+        if let Ok(bounds) = self.genv.item_bounds(opaque_def_id, span) {
+            let bounds = bounds.skip_binder();
+            let mut res = vec![];
+            for pred in &bounds.predicates {
+                if let rty::ClauseKind::Projection(pred) = pred.kind().skip_binder() {
+                    res.push(pred);
+                }
+            }
+            Some(res)
+        } else {
+            None
+        }
+    }
+
+    fn project_bty(&mut self, bty: &BaseTy, def_id: DefId) -> Ty {
+        let args = vec![GenericArg::Ty(refine_default(bty.clone()).skip_binder())];
+        let alias_ty = rty::AliasTy::new(def_id, args);
+        let proj_ty = refine_default(BaseTy::projection(alias_ty));
+        rty::projections::normalize(self.genv, self.def_id, &proj_ty, self.span())
+            .unwrap()
+            .skip_binder()
+    }
+
+    fn opaque_subtyping(&mut self, rcx: &mut RefineCtxt, bty: &BaseTy, opaque_def_id: DefId) {
+        // TODO(RJ): odd hack because a malformed predicate makes the `item_bounds` query return `Err` which cannot be `unwrapped` in `self.opaque_predicates`
```

Comment:
Ok, so I 

1. made all the `subtyping` methods now have type `Result<(), CheckerError>`
2. made `check_crate_wf` _also_ check the `DefKind::OpaqueTy` so it ends up checking the bounds for the `impl trait` in a "strict" fashion, instead of lazily ...